pkg/hiveregistrator/registrator.go

```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		return err
+	}
+
+	hr.Hiveclientset, err = hiveclient.NewForConfig(kubeConfig)
+	if err != nil {
+
+		return err
+	}
+
+	return nil
+}
+
+func (hr *hiveRegistrator) Register(ctx context.Context) error {
+	namespacename := ClusterNamespaceName(*hr.WorkloadCluster)
+	clusterdeploymentname := ClusterDeploymentName(*hr.WorkloadCluster)
+	kubesecretName := "admin-kube-secret"
+	pullsecretname := "pull-secret"
+	spsecretname := "serviceprincipal-secret"
+
+	csn := ClusterNamespace(*hr.WorkloadCluster, namespacename)
+	_, err := hr.Kubernetescli.CoreV1().Namespaces().Create(ctx, csn, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	kubesecret := KubeAdminSecret(*hr.WorkloadCluster, kubesecretName, namespacename)
+
+	_, err = hr.Kubernetescli.CoreV1().Secrets(namespacename).Create(ctx, kubesecret, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	pullsecret := PullSecret(*hr.WorkloadCluster, pullsecretname, namespacename)
+	_, err = hr.Kubernetescli.CoreV1().Secrets(namespacename).Create(ctx, pullsecret, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	spsecret := ServicePrincipalSecret(*hr.WorkloadCluster, spsecretname, namespacename)
+	_, err = hr.Kubernetescli.CoreV1().Secrets(namespacename).Create(ctx, spsecret, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	cds := ClusterDeployment(*hr.WorkloadCluster, kubesecretName, spsecretname, pullsecretname, namespacename, clusterdeploymentname)
+	_, err = hr.Hiveclientset.HiveV1().ClusterDeployments(namespacename).Create(ctx, cds, metav1.CreateOptions{})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (hr *hiveRegistrator) IsConnected(ctx context.Context) (bool, error) {
+
+	hiveNS := ClusterNamespaceName(*hr.WorkloadCluster)
+	hiveCD := ClusterDeploymentName(*hr.WorkloadCluster)
+
+	cd, err := hr.Hiveclientset.HiveV1().ClusterDeployments(hiveNS).Get(ctx, hiveCD, metav1.GetOptions{})
+	if err != nil {
+		return true, err
+	}
+
+	for _, condition := range cd.Status.Conditions {
+
+		if condition.Type == v1.UnreachableCondition {
+			//checking for false, meaning not unreachable, so is reachable
+			isReachable := condition.Status != corev1.ConditionTrue
+			return isReachable, nil
+		}
+
+	}
+
+	return true, nil
+}
+
+func (hr *hiveRegistrator) IsRegistered(ctx context.Context) (bool, error) {
+
+	hiveNS := ClusterNamespaceName(*hr.WorkloadCluster)
+	hiveCD := ClusterDeploymentName(*hr.WorkloadCluster)
+
+	ns, err := hr.Kubernetescli.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", hiveNS),
+	})
+	if err != nil {
+		return false, err
+	}
+	if len(ns.Items) == 0 {
+		//no such namespace
+		return false, nil
+	}
+
+	cds := hr.Hiveclientset.HiveV1().ClusterDeployments(hiveNS)
+	r, err := cds.List(context.TODO(), metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", hiveCD),
+	})
```

Comment:
Why do we need to list? Can't we use get here?

Reply:
That approach makes it easier to distinguish between "hive not reachable" and "resources not present"